### PR TITLE
Remove unnecessary use of ns.follow in goto-programs/

### PR DESCRIPTION
**Pull request by tautschnig**
Only structs, unions, enums can be hidden behind tags.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6eb914274dc4ca77ac8a1a7ab54ce2bfe459dca1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97032295

File: src/goto-programs/goto_convert_side_effect.cpp [hunk 1]
```diff
@@ -96,7 +96,7 @@ void goto_convertt::remove_assignment(
 
     exprt rhs;
 
-    const typet &op0_type=ns.follow(expr.op0().type());
+    const typet &op0_type = expr.op0().type();
```
Comment (smowton) [inline src/goto-programs/goto_convert_side_effect.cpp:99]: Note the test at line 108 would therefore have been unreachable prior to this change -- but evidently our tests did not reveal that. Add a test?

**Review by smowton — APPROVED** [review]

File: src/goto-programs/goto_convert_side_effect.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-programs/goto_convert_side_effect.cpp:99]: This not only was dead code, it actually still is dead code: the front-end already introduces type casts (both the C and C++ front-ends take care of this). I have, in a second commit, removed the dead code and added a `PRECONDITION` in the `else` case.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 375e6db75a3d3469a1c499ff3d2e1c2f5319519c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97056355
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0cbacbce2ab774e19896000c3aa9348b5abf7d30).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97057429

**Comment by kroening** [comment]
The first commit is definitively ok.
The second commit looks like there needs to be a fundamentally better approach to dealing with this -- it's odd that the conversion to goto would need to worry about promotion rules.

**Comment by tautschnig** [comment]
I have moved the second commit to #3770.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d4ec352674af3999c415c2ebe86e6451d6a8b465).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97132922

**Pull request merged**